### ISSUE TITLE
[REQUEST] ESM Support

### ISSUE THREAD

**Issue by giomarino** [Sparticuz/chromium#31]
ESM support is available?
Project like SST requires ESM to use all framework feature (like resource binding)

**Comment by tomchiverton** [comment]
Works out of the box, at least as an AWS Lambda, with a layer.

```
import puppeteer from "puppeteer-core";
import chromium from "@sparticuz/chromium";

export async function shot (event){
	const browser = await puppeteer.launch({
		args: chromium.args,
		defaultViewport: chromium.defaultViewport,
		executablePath: await chromium.executablePath(),
		headless: chromium.headless,
		ignoreHTTPSErrors: true,
	});

	const page = await browser.newPage();
	await page.goto("https://example.com");
	const pageTitle = await page.title();
	await browser.close();

	console.log(pageTitle, "Example Domain");
};
```

**Comment by Sparticuz** [comment]
I don't have any problems either, but the package is only set up for cjs and not esm. I'll probably eventually set it to esm only. I've heard there are problems with dual packaging

**Comment by milindsingh** [comment]
Yes I am also using it with imports. I have written an article and also made the code open source. Feel free to check it out.
https://adapttive.com/blog/serverless-scraping#how-to-debug 
https://github.com/openwishlist/preview

**Comment by bnjns** [comment]
I've just hit difficulties using this with Astro on AWS Lambda (included as a standard dependency, rather than as a layer), as I kept getting `ReferenceError: __dirname is not defined`. This appears to be coming from `chromium.executablePath`, but I've been able to get around it by providing the path to the chromium binary:

```ts
const browser = await puppeteer.launch({
  ...
    executablePath: await chromium.executablePath(`${path.dirname(fileURLToPath(import.meta.url))}/node_modules/@sparticuz/chromium/bin`),
  ...
});
```

**Comment by tomchiverton** [comment]
That variable is dropped in more recent Node versions, nothing to do with the library 
-- 
†øღ

**Comment by rdzidziguri** [comment]
Would you have any thoughts on how to fix the issue with _dirname? My code is exactly like the one above. However, I am using esbuild, and for some bizarre reason, the output is only 1.2 MB, and I wonder if it is even bundling successfully. So any advice would be appriciated.

**Comment by rdzidziguri** [comment]
Quick update, the moment I switched it to produce commonjs output, it worked without any issues, so the problem is related to the ESM bundle.

**Comment by Sparticuz** [comment]
I'm using this with typescript and esm output with a lambda layer without any issues. Generally if you are using some type of bundler is where I've seen issues.

**Comment by rdzidziguri** [comment]
esbuild is AWS recommended one. CDK uses it as well, but the moment I switch the TS config compilerOptions module to CommonJS, it works without any issues, but the moment I set it to esnext, it fails with the error above.

**Comment by dobrynin** [comment]
fwiw we are running into the same error. Thought we had it working at one point w/ ESM output, not sure if something changed or if it was actually broken the whole time.

**Comment by dobrynin** [comment]
Okay I got it working again. Not sure what the issue was but I
1. Nuked node_modules, dist directories, and cdk.out directoreis
2. tweaked our package.json (puppeteer-core went from ^19.8.1 to ^19.8.0 and @sparticuz/chromium went from ^112.0.2)
   * note that @sparticuz/chromium is just a dev dependency, we use a lambda layer in prod.
4. Updated the our chromium binary to v112.0.2

For some reason `node_modules/@puppeteer/browsers/node_modules/yargs` was removed from our package-lock.json. Our stack trace for the `__dirname` error listed `yargs` as the originating package for the error so it makes sense that the error is gone now, although I don't know why the package-lock.json changed.

**Comment by dvalbuena1** [comment]
In case someone would like to use esbuild, the only way it worked for me in AWS was as follows.

``` javascript 
import * as esbuild from "esbuild";

esbuild.build({
  entryPoints: ["out-rollup/index.js"],
  minify: true,
  bundle: true,
  platform: "node",
  format: "esm",
  target: "node18",
  external: [
    "puppeteer-core",
    "puppeteer-extra",
    "puppeteer-extra-plugin-stealth",
    "puppeteer-extra-plugin-adblocker",
    "@sparticuz/chromium",
  ],
  outdir: "dist",
  banner: {
    js: "import { createRequire } from 'module';import path from 'path';import { fileURLToPath } from 'url';const __filename = fileURLToPath(import.meta.url);const __dirname = path.dirname(__filename);const require = createRequire(import.meta.url);",
  },
  outExtension: {
    ".js": ".mjs",
  },
});
```

With the code in `banner` the error of `__dirname`, `__filename` and `require` in ESM are gone.

The downside I found is that by adding `@sparticuz/chromium` into `external` my package went from 16MB to 72MB being compressed, uncompressed was 75MB to 131MB.

**Comment by Sparticuz** [comment]
That's great info. Thanks for posting! If you are on a platform that allows you to use layers, you can use the layer, then use `@sparticuz/chromium-min`. It's the same package, just without the binaries.

**Comment by dvalbuena1** [comment]
> @sparticuz/chromium-min

You are right, I hadn't realized that!
Now the size remains at 20 MB while compressed. It is only necessary to specify the entire path using the layer.
``` javascript 
chromium.executablePath("/opt/nodejs/node_modules/@sparticuz/chromium/bin")
```

**Comment by skilbjo** [comment]
not sure if my issue is related to ESM or not. my project is in ESM, but when deploying to lambda (this is a new project), i get this as a Runtime error (ie nothing starts):

```
require() of ES Module /var/task/canary.js from /var/runtime/index.mjs not supported.\nInstead change the require of canary.js in /var/runtime/index.mjs to a dynamic import() which is available in all CommonJS module
```

i am not sure where the error emanates from, but just in case it is from the `Sparticuz/chromium` project

**Comment by Sparticuz** [comment]
A beta version has been published that adds the capability introduced in this pull request. Please try it out and provide any feedback or report issues you encounter. Your input will help ensure a smooth release.

`npm install @sparticuz/chromium@next`

or 

`npm install @sparticuz/chromium-min@next`

arm: https://github.com/Sparticuz/chromium/releases/download/v135.0.0-next.0/chromium-v135.0.0-next.0-pack.arm64.tar
x64: https://github.com/Sparticuz/chromium/releases/download/v135.0.0-next.0/chromium-v135.0.0-next.0-pack.x64.tar
